src/main/java/org/opentripplanner/updater/vehicle_parking/VehicleParkingAvailabilityUpdater.java

```diff
@@ -0,0 +1,104 @@
+import org.opentripplanner.updater.GraphWriterRunnable;
+import org.opentripplanner.updater.spi.DataSource;
+import org.opentripplanner.updater.spi.PollingGraphUpdater;
+import org.opentripplanner.updater.spi.WriteToGraphCallback;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Graph updater that dynamically sets availability information on vehicle parking lots. This
+ * updater fetches data from a single {@link DataSource<VehicleParking>}.
+ */
+public class VehicleParkingAvailabilityUpdater extends PollingGraphUpdater {
+
+  private static final Logger LOG = LoggerFactory.getLogger(
+    VehicleParkingAvailabilityUpdater.class
+  );
+  private final DataSource<AvailabiltyUpdate> source;
+  private WriteToGraphCallback saveResultOnGraph;
+
+  private final VehicleParkingService vehicleParkingService;
+
+  public VehicleParkingAvailabilityUpdater(
+    VehicleParkingUpdaterParameters parameters,
+    DataSource<AvailabiltyUpdate> source,
+    VehicleParkingService vehicleParkingService
+  ) {
+    super(parameters);
+    this.source = source;
+    this.vehicleParkingService = vehicleParkingService;
+
+    LOG.info("Creating vehicle-parking updater running every {}: {}", pollingPeriod(), source);
+  }
+
+  @Override
+  public void setup(WriteToGraphCallback writeToGraphCallback) {
+    this.saveResultOnGraph = writeToGraphCallback;
+  }
+
+  @Override
+  protected void runPolling() {
+    if (source.update()) {
+      var updates = source.getUpdates();
+
+      var graphWriterRunnable = new AvailabilityUpdater(updates);
+      saveResultOnGraph.execute(graphWriterRunnable);
+    }
+  }
+
+  private class AvailabilityUpdater implements GraphWriterRunnable {
+
+    private final List<AvailabiltyUpdate> updates;
+    private final Map<FeedScopedId, VehicleParking> parkingById;
+
+    private AvailabilityUpdater(List<AvailabiltyUpdate> updates) {
+      this.updates = List.copyOf(updates);
+      this.parkingById =
+        vehicleParkingService
+          .getVehicleParkings()
+          .collect(Collectors.toUnmodifiableMap(VehicleParking::getId, Function.identity()));
+    }
+
+    @Override
+    public void run(Graph graph, TransitModel ignored) {
+      updates.forEach(this::handleUpdate);
+    }
+
+    private void handleUpdate(AvailabiltyUpdate update) {
+      if (!parkingById.containsKey(update.vehicleParkingId())) {
+        LOG.error(
+          "Parking with id {} does not exist. Skipping availability update.",
+          update.vehicleParkingId()
+        );
+      } else {
+        var parking = parkingById.get(update.vehicleParkingId());
+        var builder = VehicleParkingSpaces.builder();
+        if (parking.hasCarPlaces()) {
+          builder.carSpaces(update.spacesAvailable());
+        }
+        if (parking.hasBicyclePlaces()) {
+          builder.bicycleSpaces(update.spacesAvailable());
+        }
+        parking.updateAvailability(builder.build());
```

Comment:
Since `VehicleParking` instances are shared between several threads and the `updateAvailability` method changes the reference to `VehicleParking.availability`, `VehicleParking.availability` should be safely published by making the field `volatile` or synchronizing all accesses to the field (This is not a new issue, this applies also to the existing code in `VehicleParkingUpdater`)
This should be documented on the field or on the `updateAvailability` method.

Reply:
Would it be enough if I added `synchronized` to the method signature?

Reply:
I've taken a stab at it here: e4199f1225f80fd399491318a4889cab7d4bed39

Reply:
You should use either volatile or synchronization, not both.
If you choose volatile, you just need to mark the field as volatile.
If you choose synchronization, then all accesses (read+write) should be synchronized, not only the write accesses.
The easiest way to enforce synchronization it is to wrap the field in a ConcurrentPublished. Then the compiler will force you to access the field through the get() and publish() methods, which are synchronized.

Reply:
What is the performance implication of a synchronized access (not write)? I'm asking because this is used during A* routing and the access should be as fast as possible.

Reply:
Probably limited impact, but this should be tested. If unsure, use volatile, since it provides safe publication without synchronization.
In my opinion the most important thing here is to document that there is concurrent access on this field which requires special handling.
The collections in `VehicleParkingService` (`vehicleParkings`, `vehicleParkingGroups`) are also swapped without ensuring safe publication, but this is out of the scope of this PR.